haskell/src/Monocle/Backend/Janitor.hs

```diff
@@ -11,6 +19,97 @@ import Monocle.Env
 import Monocle.Prelude
 import qualified Streaming.Prelude as Streaming
 
+updateAuthor :: C.Index -> D.Author -> D.Author
+updateAuthor index author@D.Author {..} = case getIdent of
+  Just ident -> D.Author ident authorUid
+  Nothing -> revertMuidIfNeeded
+  where
+    getIdent :: Maybe LText
+    getIdent = from <$> getIdentByAlias index (from authorUid)
+    revertMuidIfNeeded :: D.Author
+    revertMuidIfNeeded =
+      let muid =
+            -- Remove the host prefix
+            T.drop 1 $ T.dropWhile (/= '/') (from authorUid)
+       in if authorMuid == from muid
+            then author
+            else D.Author (from muid) authorUid
+
+-- | Apply identities according the configuration on Events
+-- Try this on the REPL with:
+-- λ> testQueryM (defaultTenant "sf-team-workspace") $ updateIdentsOnChanges
+updateIdentsOnChanges :: QueryM Int
+updateIdentsOnChanges = do
+  index <- asks tenant
+  indexName <- getIndexName
+  go indexName (updateAuthor index)
+  where
+    go indexName updateAuthor' = do
+      withQuery changeQuery $
+        scanChanges
+          & ( Streaming.mapMaybe updateChange
+                >>> Streaming.map mkEChangeBulkUpdate
+                >>> I.bulkStream
+            )
+      where
+        scanChanges :: Stream (Of D.EChange) QueryM ()
+        scanChanges = Q.scanSearchHit
+        changeQuery = mkQuery [Q.documentType D.EChangeDoc]
+        updateChange :: D.EChange -> Maybe D.EChange
+        updateChange change@D.EChange {..} =
+          let updatedChange =
+                change
+                  { echangeAuthor = updateAuthor' echangeAuthor,
+                    echangeMergedBy = updateAuthor' <$> echangeMergedBy,
+                    echangeAssignees = updateAuthor' <$> echangeAssignees,
+                    echangeCommits = updateCommitAuthors <$> echangeCommits
+                  }
+           in if updatedChange == change then Nothing else Just updatedChange
+          where
+            updateCommitAuthors :: D.Commit -> D.Commit
+            updateCommitAuthors commit@D.Commit {..} =
+              commit
+                { commitAuthor = updateAuthor' commitAuthor,
+                  commitCommitter = updateAuthor' commitCommitter
+                }
+        mkEChangeBulkUpdate :: D.EChange -> BulkOperation
+        mkEChangeBulkUpdate ec =
+          BulkUpdate indexName (I.getChangeDocId ec) $ toJSON ec
+
+-- | Apply identities according the configuration on Events
+-- Try this on the REPL with:
+-- λ> testQueryM (defaultTenant "sf-team-workspace") $ updateIdentsOnEvents
+updateIdentsOnEvents :: QueryM Int
+updateIdentsOnEvents = do
+  index <- asks tenant
+  indexName <- getIndexName
+  go indexName (updateAuthor index)
```

Comment:
ditto. Then perhaps let's make those functions always take the index name and author update function, then the caller will do the asks business.